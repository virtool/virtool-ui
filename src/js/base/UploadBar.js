import React, { useCallback } from "react";
import { useDropzone } from "react-dropzone";
import styled from "styled-components";
import { Button } from "./index";

const StyledUploadBar = styled.div`
    align-items: stretch;
    display: flex;
    height: 38px;
    margin-bottom: 15px;

    & > div:first-child {
        align-items: center;
        background-color: ${props => (props.active ? props.theme.greyLightest : "transparent")};
        border: 1px solid ${props => (props.active ? props.theme.blue : props.theme.color.greyLight)};
        border-radius: ${props => props.theme.borderRadius.sm};
        box-sizing: border-box;
        cursor: pointer;
        display: flex;
        flex: 1 0 auto;
        justify-content: center;
        margin: 0;
    }

    & > button {
        flex: 0 0 auto;
        margin-left: 3px;
    }
`;

export const UploadBar = ({ message, onDrop }) => {
    const messageComponent = <span>{message && message.length ? message : "Drag file here to upload"}</span>;

    const handleDrop = useCallback(acceptedFiles => {
        onDrop(acceptedFiles);
    }, []);

    const { getRootProps, getInputProps, isDragAccept, open } = useDropzone({ onDrop: handleDrop });

    const rootProps = getRootProps({
        onClick: e => e.stopPropagation()
    });

    return (
        <StyledUploadBar active={isDragAccept}>
            <div {...rootProps}>
                <input {...getInputProps()} />
                {messageComponent}
            </div>
            <Button color="blue" icon="upload" onClick={open}>
                Upload
            </Button>
        </StyledUploadBar>
    );
};
