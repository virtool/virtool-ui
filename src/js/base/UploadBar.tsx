import React, { useCallback } from "react";
import { FileError, useDropzone } from "react-dropzone";
import styled, { DefaultTheme } from "styled-components";
import { getColor } from "../app/theme";
import { Button } from "./Button";
import { DividerVertical } from "./DividerVertical";

type getUploadBarColorProps = {
    /* Whether the user is able to drag/select files */
    active: boolean;
    theme: DefaultTheme;
};

function getUploadBarBackgroundColor({ active, theme }: getUploadBarColorProps): string {
    return active ? getColor({ color: "greyLightest", theme }) : "transparent";
}

function getUploadBarBorderColor({ active, theme }: getUploadBarColorProps): string {
    return getColor({ theme, color: active ? "blue" : "greyLight" });
}

type StyledUploadBarProps = {
    /* Whether the user is able to drag/select files */
    active: boolean;
};

const StyledUploadBar = styled.div<StyledUploadBarProps>`
    display: flex;
    align-items: stretch;
    justify-content: center;
    padding: 10px 15px;
    margin-bottom: 15px;

    background-color: ${getUploadBarBackgroundColor};
    border: 1px solid ${getUploadBarBorderColor};
    border-radius: ${props => props.theme.borderRadius.sm};
    cursor: pointer;

    button {
        margin: auto 0;
    }
`;

const MessageContainer = styled.div`
    display: flex;
    flex: 1 0 50px;
    justify-content: end;
    align-items: center;
    min-height: 60px;
    margin: auto 0;
`;

const ButtonContainer = styled.div`
    display: flex;
    flex: 1 0 50px;
`;

const UploadBarDivider = styled(DividerVertical)`
    margin: 0 40px;

    span {
        text-transform: uppercase;
    }
`;

type UploadBarProps = {
    message?: React.ReactNode;
    /* Whether multiple files can be uploaded */
    multiple?: boolean;
    /* Callback when the upload bar loses focus */
    onBlur?: () => void;
    /* Callback when files are dropped */
    onDrop: (acceptedFiles: File[]) => void;
    /* Validates if a file is allowed */
    validator?: (file: File) => FileError;
};

/*
 * Allows files to be dragged and dropped or selected from the file system.
 */
export function UploadBar({
    message = "Drag file here to upload",
    multiple = true,
    onBlur,
    onDrop,
    validator,
}: UploadBarProps) {
    const handleDrop = useCallback(
        (acceptedFiles: File[]) => {
            onDrop(acceptedFiles);
        },
        [onDrop],
    );

    const { getRootProps, getInputProps, isDragAccept, open } = useDropzone({ onDrop: handleDrop, validator });

    const rootProps = getRootProps({
        onClick: e => e.stopPropagation(),
    });

    return (
        <StyledUploadBar active={isDragAccept} {...rootProps}>
            <input {...getInputProps()} aria-label="Upload file" multiple={multiple} />
            <MessageContainer>{message}</MessageContainer>
            <UploadBarDivider text="or" />
            <ButtonContainer>
                <Button color="blue" icon="upload" onClick={open} onBlur={onBlur}>
                    Browse Files
                </Button>
            </ButtonContainer>
        </StyledUploadBar>
    );
}
