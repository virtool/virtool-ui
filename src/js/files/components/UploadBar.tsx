import { Button } from "@base/Button";
import { Icon } from "@base/Icon";
import { cn } from "@utils/utils";
import React, { useCallback } from "react";
import { Accept, FileError, useDropzone } from "react-dropzone";

type UploadBarProps = {
    accept?: Accept;

    /** The message to display in the upload bar */
    message?: React.ReactNode;

    /** Whether multiple files can be uploaded */
    multiple?: boolean;

    /** Callback when the upload bar loses focus */
    onBlur?: () => void;

    /** Callback when files are dropped */
    onDrop: (acceptedFiles: File[]) => void;

    /** A regular expression to validate file names */
    regex?: RegExp;
};

/*
 * Allows files to be dragged and dropped or selected from the file system.
 */
export function UploadBar({
    accept,
    message = "Drag file here to upload",
    multiple = true,
    onBlur,
    onDrop,
    regex,
}: UploadBarProps) {
    const handleDrop = useCallback(
        (acceptedFiles: File[]) => {
            onDrop(acceptedFiles);
        },
        [onDrop]
    );

    function validator(file: File): FileError {
        if (regex && !regex.test(file.name)) {
            return {
                code: "file-invalid-type",
                message: "Invalid file type",
            };
        }
    }

    const { fileRejections, getRootProps, getInputProps, isDragAccept, isDragReject, open } = useDropzone({
        accept,
        onDrop: handleDrop,
        validator,
    });

    const rootProps = getRootProps({
        onClick: e => e.stopPropagation(),
    });

    return (
        <div
            className={cn(
                {
                    "bg-zinc-100": isDragAccept && !isDragReject,
                    "bg-zinc-50": !isDragAccept && !isDragReject,
                    "bg-red-100": isDragReject && !isDragAccept,
                },
                "border",
                {
                    "border-solid": !isDragReject || !isDragAccept,
                    "border-dashed": isDragReject || isDragAccept,
                },
                {
                    "border-slate-400": !isDragReject,
                    "border-red-400": isDragReject,
                },
                "flex",
                "flex-col",
                "items-center",
                "mb-4",
                "p-2",
                "pt-10",
                "rounded-md"
            )}
            {...rootProps}
        >
            <input {...getInputProps()} aria-label="Upload file" multiple={multiple} />
            <div className="gap-2 grid grid-cols-11 place-items-stretch">
                <div className="col-span-5 flex items-center justify-end">{message}</div>
                <div className="col-span-1 flex font-bold items-center justify-center text-gray-600">OR</div>
                <div className="col-span-5 flex items-center justify-start">
                    <Button color="blue" onBlur={onBlur} onClick={open}>
                        <Icon name="upload" /> Browse Files
                    </Button>
                </div>
            </div>
            <div className="font-medium h-6 mt-4 text-red-600 text-sm">
                {fileRejections.length > 0 &&
                    `Invalid file names: ${fileRejections.map(({ file }) => file.name).join(", ")}`}
            </div>
        </div>
    );
}
