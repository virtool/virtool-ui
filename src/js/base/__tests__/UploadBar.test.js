import { UploadBar } from "../UploadBar";

describe("<UploadBar />", () => {
    it("should render", () => {
        const wrapper = shallow(<UploadBar />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should render optional message", () => {
        const wrapper = shallow(<UploadBar message="Foobar" />);
        expect(wrapper).toMatchSnapshot();
    });
});
