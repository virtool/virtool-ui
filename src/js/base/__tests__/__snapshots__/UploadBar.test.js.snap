// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UploadBar /> should render 1`] = `
<UploadBar__StyledUploadBar
  active={false}
>
  <div
    onBlur={[Function]}
    onClick={[Function]}
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    tabIndex={0}
  >
    <input
      autoComplete="off"
      multiple={true}
      onChange={[Function]}
      onClick={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      tabIndex={-1}
      type="file"
    />
    <span>
      Drag file here to upload
    </span>
  </div>
  <Button
    color="blue"
    disabled={false}
    icon="upload"
    onClick={[Function]}
    tipPlacement="top"
    type="button"
  >
    Upload
  </Button>
</UploadBar__StyledUploadBar>
`;

exports[`<UploadBar /> should render optional message 1`] = `
<UploadBar__StyledUploadBar
  active={false}
>
  <div
    onBlur={[Function]}
    onClick={[Function]}
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    tabIndex={0}
  >
    <input
      autoComplete="off"
      multiple={true}
      onChange={[Function]}
      onClick={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      tabIndex={-1}
      type="file"
    />
    <span>
      Foobar
    </span>
  </div>
  <Button
    color="blue"
    disabled={false}
    icon="upload"
    onClick={[Function]}
    tipPlacement="top"
    type="button"
  >
    Upload
  </Button>
</UploadBar__StyledUploadBar>
`;
